### PR TITLE
Usability Improvements

### PR DESCRIPTION
**Pull request by kferrone**
### **User description**
Added devcontainers and vscode setup.

___

### **PR Type**
Enhancement, Documentation, Tests, Configuration changes

___

### **Description**
- Enhanced MkDocs script with new filters and static file handling.
- Added detailed docstrings and `apply` method for services.
- Introduced command aliases for tenant and user resources.
- Enhanced argument handling and added detailed docstrings.
- Added setup script for environment preparation.
- Added devcontainer configuration for VS Code.
- Added Dependabot configuration for devcontainers.
- Added workflow for auditing dependencies and quality checks.
- Updated Homebrew and image build workflows.
- Added recommended extensions, launch configurations, settings, and tasks for VS Code.
- Updated changelog and contributing guidelines.
- Added devcontainer build stage and updated build stages in Dockerfile.
- Updated docker-compose configuration for multiarch builds and added devcontainer service.
- Updated MkDocs configuration with new navigation and options.
- Added `pip-audit` to test dependencies.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>10 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>mkdocs.py</strong><dd><code>Enhance MkDocs script with new filters and static file handling</code></dd></summary>
<hr>

scripts/mkdocs.py

<li>Added new filters for CLI arguments and string handling.<br> <li> Introduced <code>copy_static</code> function for copying static files.<br> <li> Enhanced <code>on_startup</code> and <code>on_config</code> functions to use <code>copy_static</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/96/files#diff-53127e811461856062258e6a5adf882eb5855106bc414a7554ec8228f53d9d63">+21/-1</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>service.py</strong><dd><code>Add detailed docstrings and apply method for services</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/duplo_resource/service.py

<li>Added detailed docstrings for service methods.<br> <li> Updated methods to return dictionaries.<br> <li> Introduced <code>apply</code> method for services.<br>

</details>

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/96/files#diff-de4bdab4d945f3f4c737be6a44e3c3c11850f302766a911d4bc30229766398ab">+94/-20</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>tenant.py</strong><dd><code>Add command aliases for tenant resource</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/duplo_resource/tenant.py

- Added aliases for `list` and `find` commands.

</details>

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/96/files#diff-7065eff8535e2d18fbe9f8011540cf4f22ac22c967e8ebc19e3a9c1601f23419">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>user.py</strong><dd><code>Add command aliases for user resource</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/duplo_resource/user.py

- Added aliases for `list` and `find` commands.

</details>

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/96/files#diff-1db883fb1de60ca53d2da0d3d774f6121013f7ef1f7fe87b575a03b275a602f4">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>args.py</strong><dd><code>Enhance argument handling and add detailed docstrings</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/duplocloud/args.py

<li>Added detailed docstrings for argument definitions.<br> <li> Introduced <code>available_formats</code> function.<br> <li> Enhanced argument handling and default value retrieval.<br>

</details>

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/96/files#diff-10408bee68a44ad34e4226ec91f20e3277d2d637fd8e17d3482fea817838ec96">+40/-15</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>argtype.py</strong><dd><code>Refactor argument handling and add type name property</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/duplocloud/argtype.py

<li>Refactored argument handling to improve default value retrieval.<br> <li> Added <code>type_name</code> property for argument type names.<br>

</details>

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/96/files#diff-049d08f3ed7efa161eb37e5097bb2b1c7450bc728a16e9c3e7e89d4a81590a07">+19/-7</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>client.py</strong><dd><code>Update client initialization to use extracted arguments</code>&nbsp; &nbsp; </dd></summary>
<hr>

src/duplocloud/client.py

- Updated `from_env` and `from_args` to use `extract_args`.

</details>

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/96/files#diff-ed921a2441e8df6f226f7024b875dea79afcd0d87e1326c37b19d232dba3f740">+5/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>commander.py</strong><dd><code>Add command alias support and enhance command parsing</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/duplocloud/commander.py

<li>Added support for command aliases.<br> <li> Introduced <code>extract_args</code> and <code>aliased_method</code> functions.<br> <li> Enhanced command registration and parsing.<br>

</details>

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/96/files#diff-76bc85ae5e9281fbbae1c51d5775fc1663f2b323368722e8bf16a21578a9b9f1">+52/-22</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>resource.py</strong><dd><code>Add apply method and enhance command handling for resources</code></dd></summary>
<hr>

src/duplocloud/resource.py

<li>Introduced <code>apply</code> method for resources.<br> <li> Enhanced command handling with argument extraction.<br>

</details>

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/96/files#diff-9826d086af2db3060ef4f14b0bfd46d14eca8da014a08d6e1461a6794075c5f0">+34/-7</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>setup.sh</strong><dd><code>Add setup script for environment preparation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

scripts/setup.sh

<li>Added setup script for creating directories and installing <br>dependencies.<br>

</details>

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/96/files#diff-ce435f38623d032d25670f2c63a430dfd1a02db661d10c1ef031cbef2f0f0ef0">+5/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>test_commander.py</strong><dd><code>Add tests for argument extraction and command parsing</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/tests/test_commander.py

- Added tests for argument extraction and command parsing.

</details>

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/96/files#diff-d5e0ae6b2eca2d4bd76d259d189fab0a133f3ffaeecbf216e0b1dd6ffc47db74">+14/-14</a>&nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Configuration changes</strong></td><td><details><summary>11 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>.devcontainer.json</strong><dd><code>Add devcontainer configuration for VS Code</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

.devcontainer.json

- Added devcontainer configuration for VS Code.

</details>

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/96/files#diff-8c75d30eaaa096649631dde76e321d13a6849c64137f3f7f777429ef9fd6b9e2">+51/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>dependabot.yml</strong><dd><code>Add Dependabot configuration for devcontainers</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

.github/dependabot.yml

- Added Dependabot configuration for devcontainers.

</details>

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/96/files#diff-dd4fbda47e51f1e35defb9275a9cd9c212ecde0b870cba89ddaaae65c5f3cd28">+12/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>audit.yml</strong><dd><code>Add audit and quality workflow</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

.github/workflows/audit.yml

- Added workflow for auditing dependencies and quality checks.

</details>

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/96/files#diff-93a89f16f6717d1e4e4e27ec20ef916e3355c7aa890ca74cd9b12db30d8a899f">+68/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>homebrew.yml</strong><dd><code>Update Homebrew workflow setup action</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

.github/workflows/homebrew.yml

- Updated setup action to use editable install.

</details>

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/96/files#diff-bef3099379a9a034a0ee7c4eb0e6eb98cde146b03245e3cdb27157c36062c400">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>image.yml</strong><dd><code>Update image build workflow</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

.github/workflows/image.yml

- Updated build-image action reference.

</details>

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/96/files#diff-113df80dfdb383808d66c98190c3f60ea45745427f926da3b04f572ff081d8a6">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>extensions.json</strong><dd><code>Add recommended extensions for VS Code</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

.vscode/extensions.json

- Added recommended extensions for VS Code.

</details>

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/96/files#diff-c16655a98a3ee89a7636a59c59a72b0e93649e3a1e947327cfc43a1336b4e912">+12/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>launch.json</strong><dd><code>Add launch configurations for debugging and testing</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

.vscode/launch.json

- Added launch configurations for debugging and testing.

</details>

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/96/files#diff-bd5430ee7c51dc892a67b3f2829d1f5b6d223f0fd48b82322cfd45baf9f5e945">+51/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>settings.json</strong><dd><code>Add VS Code settings for Python and editor customization</code>&nbsp; </dd></summary>
<hr>

.vscode/settings.json

- Added VS Code settings for Python and editor customization.

</details>

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/96/files#diff-a5de3e5871ffcc383a2294845bd3df25d3eeff6c29ad46e3a396577c413bf357">+20/-0</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>tasks.json</strong><dd><code>Add tasks for building, auditing, and testing</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

.vscode/tasks.json

- Added tasks for building, auditing, and testing.

</details>

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/96/files#diff-7d76d7533653c23b753fc7ce638cf64bdb5e419927d276af836d3a03fdf1745a">+146/-0</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>Dockerfile</strong><dd><code>Add devcontainer build stage and update build stages</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

Dockerfile

<li>Added devcontainer build stage.<br> <li> Updated build stages for package and binaries.<br>

</details>

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/96/files#diff-dd2c0eb6ea5cfc6c4bd4eac30934e2d5746747af48fef6da689e85b752f39557">+11/-3</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>docker-compose.yaml</strong><dd><code>Update docker-compose for multiarch builds and add devcontainer </code><br><code>service</code></dd></summary>
<hr>

docker-compose.yaml

<li>Updated docker-compose configuration for multiarch builds.<br> <li> Added devcontainer service.<br>

</details>

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/96/files#diff-3fde9d1a396e140fefc7676e1bd237d67b6864552b6f45af1ebcc27bcd0bb6e9">+27/-8</a>&nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Documentation</strong></td><td><details><summary>3 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>CHANGELOG.md</strong><dd><code>Update changelog with new features and improvements</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

CHANGELOG.md

- Updated changelog with new features and improvements.

</details>

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/96/files#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4ed">+11/-1</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>CONTRIBUTING.md</strong><dd><code>Expand contributing guidelines with detailed setup instructions</code></dd></summary>
<hr>

CONTRIBUTING.md

- Expanded contributing guidelines with detailed setup instructions.

</details>

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/96/files#diff-eca12c0a30e25b4b46522ebf89465a03ba72a03f540796c979137931d8f92055">+59/-22</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>mkdocs.yml</strong><dd><code>Update MkDocs configuration with new navigation and options</code></dd></summary>
<hr>

mkdocs.yml

- Updated MkDocs configuration with new navigation and options.

</details>

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/96/files#diff-98d0f806abc9af24e6a7c545d3d77e8f9ad57643e27211d7a7b896113e420ed2">+2/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Dependencies</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>pyproject.toml</strong><dd><code>Add pip-audit to test dependencies</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pyproject.toml

- Added `pip-audit` to test dependencies.

</details>

  </td>
  <td><a href="https://github.com/duplocloud/duploctl/pull/96/files#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -21,3 +21,8 @@ venv
 .github/env
 requirements.txt
 actions-runner/
+stuff/
```
Comment (duplocloud-matt) [inline .gitignore:24]: Is this an example or something left in automatically?

File: .vscode/settings.json [hunk 2]
```diff
@@ -0,0 +1,20 @@
+{
+  "task.autoDetect": "off",
+  "python.terminal.activateEnvInCurrentTerminal": true,
+  "python.testing.pytestEnabled": true,
+  "python.testing.pytestArgs": [
+    "src",
+    "--no-cov",
+    "-s"
+  ],
+  "workbench.iconTheme": "bearded-icons",
```
Comment (duplocloud-matt) [inline .vscode/settings.json:10]: Do we want this included?  I'm not sure what it does but it seems odd to include themes with the project.

Comment (kferrone) [reply .vscode/settings.json]: I actually intended to skip that. 
I ran this command so subsequent updates are ignored. 
```sh
git update-index --skip-worktree .vscode/settings.json
```
This way, if you want you can simply remove those from your settings and git will completely ignore any changes you make to that file.

File: .gitignore (hunk 1 above)
Comment (kferrone) [reply .gitignore:24]: No, I have `stuff` and `foo.*` ignored globally on my machine. Using the new devcontainer locally mounts the entire working dir which doesn't have my global ignore. Therefore a few hundred files popped up to be commited. So I had to ignore these in the projects gitignore.

**Pull request merged**